### ISSUE TITLE
Timeouts not implemented

### ISSUE THREAD

**Issue by petercrowley** [cosullivan/SmtpServer#48]
I can't find any information on how to set any time outs.  I have a bad client that consistently sends the DATA command, then doesn't send anything after that.  The server just sits and waits forever, wasting resources.  I have to reset it manually.  I'd like to be able to set it to so many milliseconds, say 2 minutes worth, and then the system sends a 501 and closes the session.

https://tools.ietf.org/html/rfc2821#section-4.5.3.2

See RFC 2821 Section 4.5.3.2

An SMTP client MUST provide a timeout mechanism.  It MUST use per-
   command timeouts rather than somehow trying to time the entire mail
   transaction.  Timeouts SHOULD be easily reconfigurable, preferably
   without recompiling the SMTP code.  To implement this, a timer is set
  for each SMTP command and for each buffer of the data transfer.  The
   latter means that the overall timeout is inherently proportional to
   the size of the message.

**Comment by cosullivan** [comment]
Hi,

At the moment there is no way to set the timeout, it relies on the session being terminated by the client. It is on my list to add support for this though, so I will do that in the coming days.

The reference to the timeouts that you list in your message are actually referring to the timeout the client should apply whilst waiting for the response from the server.

Thanks,
Cain.

**Comment by petercrowley** [comment]
Thank you Cain.  I didn't realize that was a client document.  I'm anticipating the new code with happiness.

**Comment by petercrowley** [comment]
Here's what I came up with.  Use it if you want.

Additional TaskExtensions:
        public static async Task<T> TimeoutAfter<T>(this Task<T> task, int millisecondsTimeout, CancellationToken cancellationToken)
        {
            if (task == await Task.WhenAny(task, Task.Delay(millisecondsTimeout)))
                return task.Result;
            else
                throw new OperationCanceledException("Timeout Reached", new TimeoutException(), cancellationToken);
        }

        public static async Task TimeoutAfter(this Task task, int millisecondsTimeout, CancellationToken cancellationToken)
        {
            if (task == await Task.WhenAny(task, Task.Delay(millisecondsTimeout)))
                await task;
            else
                throw new OperationCanceledException("Timeout Reached", new TimeoutException(), cancellationToken);
        }

For a Read Timeout, I added in SMTPSession:
                       try
{ var text = await context.Client.ReadLineAsync(cancellationToken).TimeoutAfter(readTimeout, cancellationToken).ReturnOnAnyThread();
                    catch (OperationCanceledException cancelException)
                    {
                        context.IsQuitRequested = true;
                        await context.Client.ReplyAsync(new SmtpResponse(SmtpReplyCode.ServiceClosingTransmissionChannel, "Timeout Reached"), cancellationToken).ConfigureAwait(false);
                        //_cts.Cancel();
                    }
For the Commands, I added this: Example DataCommand:
            context.Transaction.Message = await ReadMessageAsync(context, cancellationToken).TimeoutAfter(1000 * 2, cancellationToken).ConfigureAwait(false);

**Comment by cosullivan** [comment]
Hi, thanks for your input.

I have added support for these timeouts now in a pre-release version;
https://www.nuget.org/packages/SmtpServer/4.3.0-Beta2

I have also exposed the timeout's on the OptionBuilder as;

    /// <summary>
    /// The timeout to use when waiting for a command from the client.
    /// </summary>
    TimeSpan CommandWaitTimeout { get; }

    /// <summary>
    /// The timeout on each individual buffer read.
    /// </summary>
    TimeSpan NetworkBufferReadTimeout { get; }

I haven't added a timeout for the entire message, but rather for each individual buffer read operation (current the buffer is read at 128 bytes, but I've also added this via the options builder too).

Thanks,
Cain.
